### PR TITLE
Allow saving screenshots as JPGs

### PR DESCRIPTION
**Pull request by Vultraz**
See commit messages for more details.

### CODE REVIEW

File: src/image.cpp [hunk 1]
```diff
@@ -1335,15 +1331,12 @@ bool save_image(const surface& surf, const std::string& filename)
 		return false;
 	}
 
-#ifdef HAVE_LIBPNG
 	if(!filesystem::ends_with(filename, ".bmp")) {
 		LOG_DP << "Writing a png image to " << filename << std::endl;
 
-		surface tmp = SDL_PNGFormatAlpha(surf.get());
-		const int err = SDL_SavePNG_RW(tmp, filesystem::make_write_RWops(filename).release(), 1); //1 means SDL takes ownership of the RWops
+		const int err = IMG_SavePNG_RW(surf, filesystem::make_write_RWops(filename).release(), true); // SDL takes ownership of the RWops
```
Comment (jyrkive) [inline src/image.cpp]: The last parameter is an int, so please pass 1 instead of true.

File: src/image.cpp [hunk 2]
```diff
@@ -1331,7 +1331,16 @@ bool save_image(const surface& surf, const std::string& filename)
 		return false;
 	}
 
-	if(!filesystem::ends_with(filename, ".bmp")) {
+#if SDL_IMAGE_VERSION_ATLEAST(2, 0, 2)
+	if(filesystem::ends_with(filename, ".jpg")) {
+		LOG_DP << "Writing a png image to " << filename << std::endl;
```
Comment (jyrkive) [inline src/image.cpp]: Copy-paste mistake. This line should say that it's writing a JPEG image.

(BTW, I wonder why these log messages are writing acronyms in lowercase.)

File: src/image.cpp [hunk 3]
```diff
@@ -1331,7 +1331,16 @@ bool save_image(const surface& surf, const std::string& filename)
...
+
+		const int err = IMG_SaveJPG_RW(surf, filesystem::make_write_RWops(filename).release(), true, 50); // SDL takes ownership of the RWops
```
Comment (jyrkive) [inline src/image.cpp]: Same here: the third parameter is an int, so please pass 1.

**Comment by CelticMinstrel** [comment]
Um, why was JPEG made the default? PNG is preferable in my opinion.

**Comment by Vultraz** [comment]
@CelticMinstrel 7MB -> ~500 KB file size is enough reason, I think :P

**Comment by jyrkive** [comment]
@CelticMinstrel See @ancestral 's test in 2015: https://forums.wesnoth.org/viewtopic.php?p=580511&sid=cdf76f5de7808da62419e9dbff9a746c#p580511

> I took a screenshot of the first map in _Two Brothers_. I “took the screenshot” in the game to save the map file, then created the image outside the game through Wesnoth via the command line. It created a 14.5 MB BMP file. I saved a copy as PNG at 9 MB and optimized it down 8.5 MB. Meanwhile, I saved a copy as a JPG at 50% quality at 1.3 MB and I can _easily_ say it still looks great.

**Comment by CelticMinstrel** [comment]
Can we at add a preference for it at least? It would have to be a hidden preference though since we're on strong freeze. (Like my "depreciations to chat" preference.)

**Comment by jyrkive** [comment]
As Vultraz said, the plan is to add a preference later.

**Comment by AI0867** [comment]
JPEG is lossy, especially at quality 50. I don't consider this to be an acceptable default. (at quality 100, it might be)

**Comment by Vultraz** [comment]
It looked fine when I tested. Though I was looking at map screenshots (full map, zoomed out) and not the screenshots that are just of the fish lemons viewport... hm.

**Comment by AI0867** [comment]
@jyrkive look at the next page in that thread: https://forums.wesnoth.org/viewtopic.php?p=580524#p580524

Sure, the map comes through fine, but as a background, it's intentionally lacking in high-frequency data. Units (proper pixel art) fare less well, and hard-edged bits of the interface (text, status bars) show clear jpeg-artifacts. I can still distinguish these issues at 80% quality. 50% is just appalling.

**Comment by jyrkive** [comment]
Ah. I hadn't read the second page.

Unfortunately, IMG_SaveJPG_RW() doesn't offer an option to disable chroma subsampling. :(

I guess that the decision to implement saving as JPEGs was done too hastily in order to stay within our overly-tight schedule... (Not that I'm fully free of the blame, either, since this PR carries my approval.)

**Comment by CelticMinstrel** [comment]
Just add the preference (hidden, in that the only way to set it is to edit your preferences file) and put the default back to png?

**Comment by jyrkive** [comment]
Yes, sounds like a good option.

**Comment by Vultraz** [comment]
IMO, one should consider just what purpose most people would want their screenshots for. Do more people want pixel-perfect reproductions to frame and hang on their wall, or do more people want quick screenshots to share on Discord and the like?

Either way, I think a better solution is this: instead of merely a hidden preference, or even a visible one, we should instead display the screenshot prompt *before* saving the screenshot, and have a radio button group for format. Right now it doesn't display until *after* the shot is saved. This would allow people to specify filenames too. This could still be accompanied by a preference to say which one was selected by default.

**Comment by CelticMinstrel** [comment]
But that can't be done until 1.15, because strong freeze. A hidden preference could be done right now.

**Comment by Vultraz** [comment]
The strings could be marked untranslatable.

**Comment by CelticMinstrel** [comment]
That's not a great solution for clearly user-visible strings.

**Comment by Vultraz** [comment]
It can be made translatable for 1.14.1.

**Comment by CelticMinstrel** [comment]
That still leaves a non-translatable string visible for 1.14.1.

**Comment by CelticMinstrel** [comment]
Obviously I meant 1.14.0.

**Comment by AI0867** [comment]
I would prefer the hidden preference option for 1.14.0 then.

As a workaround, we could also increase the jpeg quality to reduce artifacts.

**Comment by ancestral** [comment]
@AI0867 It really sounds like you are not the target audience of JPEG screenshots, which is fine. JPEG is *intended* to be smaller in size, and graphical quality is *intended* to be less than perfect. Depending on what the file savings are, I’m not sure it’s worthwhile to have the quality at or near 100. But, if we don’t know what value to use, the thing to do is try it at different values, compare file sizes, compare perceived quality, and choose something. (50, 75. 90, 100?)

As far as whether JPEG or PNG is default and what is not, I don’t really care myself, as long as the option is there to change between the formats. (To be honest, the more I think about it, the more I like PNG as the default, too, but I see the value of changing it as JPEG for many users.)

**Comment by AI0867** [comment]
With quality 100, JPEG are still both quite a bit smaller in size and somewhat lossy. For this reason, JPEG is used for story screen images, at a high quality factor (somewhere in the range of 90-100, IIRC). Given the hard edges in the interface that screenshots will also have to deal with, that would be the absolute minimum if you want to consider making it the default.

If it's not the default, it's less important of course, but you may instead want to place the quality factor under user control. (so the user can pick from the range of "I want to save half (random guess) my bytes but it still needs to look good" to "I don't care about quality, I just want to show people this situation on the map, and I'm on 56k")

**Comment by ancestral** [comment]
I can’t agree with you without empirical evidence using Wesnoth screenshots.

**Comment by jyrkive** [comment]
@Vultraz I have now implemented showing the screenshot prompt before saving the screenshot. I also changed the default screenshot format back to PNG.

**Pull request merged**